source/mesh_deformation/interface.cc

```diff
@@ -722,6 +740,253 @@ namespace aspect
+           ++level)
+        {
+          if (level > 0)
+            {
+              smoother_data[level].smoothing_range = 15.;
+              smoother_data[level].degree = 5;
+              smoother_data[level].eig_cg_n_iterations = 10;
+            }
+          else
+            {
+              smoother_data[0].smoothing_range = 1e-3;
+              smoother_data[0].degree = numbers::invalid_unsigned_int;
+              smoother_data[0].eig_cg_n_iterations = mg_matrices[0].m();
+            }
+          mg_matrices[level].compute_diagonal();
+          smoother_data[level].preconditioner =
+            mg_matrices[level].get_matrix_diagonal_inverse();
+        }
+      mg_smoother.initialize(mg_matrices, smoother_data);
+      MGCoarseGridApplySmoother<dealii::LinearAlgebra::distributed::Vector<double>> mg_coarse;
+      mg_coarse.initialize(mg_smoother);
+
+      // set up the interface matrices
+      mg::Matrix<dealii::LinearAlgebra::distributed::Vector<double>> mg_matrix(mg_matrices);
+      MGLevelObject<MatrixFreeOperators::MGInterfaceOperator<SystemMatrixType>> mg_interface_matrices;
+      mg_interface_matrices.resize(0, n_levels - 1);
+      for (unsigned int level = 0; level < n_levels;
+           ++level)
+        mg_interface_matrices[level].initialize(mg_matrices[level]);
+      mg::Matrix<dealii::LinearAlgebra::distributed::Vector<double>> mg_interface(mg_interface_matrices);
+      Multigrid<dealii::LinearAlgebra::distributed::Vector<double>> mg(mg_matrix, mg_coarse, mg_transfer, mg_smoother, mg_smoother);
+      mg.set_edge_matrices(mg_interface, mg_interface);
+      PreconditionMG<dim,
+                     dealii::LinearAlgebra::distributed::Vector<double>,
+                     MGTransferMatrixFree<dim, double>>
+                     preconditioner(mesh_deformation_dof_handler, mg, mg_transfer);
+
+      // solve
+      const double tolerance
+        = sim.parameters.linear_stokes_solver_tolerance
+          * ((this->simulator_is_past_initialization()) ? 1.0 : 1e-5);
+
+      SolverControl solver_control_mf(5 * corrected_rhs.size(),
+                                      tolerance * corrected_rhs.l2_norm());
+      SolverCG<dealii::LinearAlgebra::distributed::Vector<double>> cg(solver_control_mf);
+
+      mesh_velocity_constraints.set_zero(mf_solution);
+      cg.solve(mf_mesh_matrix, mf_solution, corrected_rhs, preconditioner);
+      this->get_pcout() << "   Solving mesh displacement system using GMG... " << solver_control_mf.last_step() <<" iterations."<< std::endl;
+
+      mesh_velocity_constraints.distribute(mf_solution);
+      mf_solution.update_ghost_values();
+
+      // copy solution:
+      LinearAlgebra::Vector solution_tmp;
+      solution_tmp.reinit(mesh_locally_owned, sim.mpi_communicator);
+      dealii::LinearAlgebra::ReadWriteVector<double> rwv(solution_tmp.locally_owned_elements());
+      rwv.import(mf_solution, VectorOperation::insert);
+      solution_tmp.import(rwv, VectorOperation::insert);
```

Comment:
Can you extend the comment above to why this is so complicated? I know it is because of the conversion ghosted/non-ghosted vectors, but it seems surprising we need to make that many copies just make the assignment `mf_solution` -> `fs_mesh_velocity`.

Reply:
Is there an easier way to do this? @tjhei